src/Filesystem.cpp

```diff
@@ -417,29 +322,178 @@ std::string Filesystem::workingPath(const std::string& filename) const
+}
+
+bool Filesystem::isInit() const
+{
+    return mIsInit;
+}
+
+bool Filesystem::exists(const std::filesystem::path& p) const
+{
+    return std::filesystem::exists(p);
+}
+
+void Filesystem::forEachFileInFolder(const std::filesystem::path& folderpath, const std::string& validExtensionList /*= std::string{}*/, const std::function<void(const std::filesystem::path&)>& callback /*= [](const std::filesystem::path& p) { (void)p; }*/, bool recursive /*= false*/) const {
+    namespace FS = std::filesystem;
+    auto preferred_folderpath = folderpath;
+    preferred_folderpath.make_preferred();
+    bool exists = FS::exists(preferred_folderpath);
+    bool is_directory = FS::is_directory(preferred_folderpath);
+    bool is_folder = exists && is_directory;
+    if(!is_folder) {
+        return;
+    }
+    auto validExtensions = split(toLowercase(validExtensionList));
+    if(!recursive) {
+        forEachFileInFolders<FS::directory_iterator>(preferred_folderpath, validExtensions, callback);
+    } else {
+        forEachFileInFolders<FS::recursive_directory_iterator>(preferred_folderpath, validExtensions, callback);
+    }
+}
+
+void Filesystem::forEachFileInFolder(const std::filesystem::path& folderpath, const std::string& validExtensionList /*= std::string{}*/, const std::function<void(const std::filesystem::path&)>& callback /*= [](const std::filesystem::path& p) { (void)p; }*/, bool recursive /*= false*/) {
+    static_cast<const Filesystem&>(*this).forEachFileInFolder(folderpath, validExtensionList, callback, recursive);
+}
+
+bool Filesystem::readBufferFromFile(std::vector<unsigned char>& out_buffer, const std::string& filePath) const {
+    namespace FS = std::filesystem;
+    FS::path p(filePath);
+    p.make_preferred();
+    bool path_is_directory = FS::is_directory(p);
+    bool path_not_exist = !FS::exists(p);
+    bool not_valid_path = path_is_directory || path_not_exist;
+    if(not_valid_path) {
+        return false;
+    }
+
+    auto byte_size = static_cast<std::size_t>(FS::file_size(p));
+    out_buffer.resize(byte_size);
+    std::ifstream ifs;
+    ifs.open(p, std::ios_base::binary);
+    ifs.read(reinterpret_cast<char*>(out_buffer.data()), out_buffer.size());
+    ifs.close();
+    out_buffer.shrink_to_fit();
+    return true;
+
+}
+
+bool Filesystem::readBufferFromFile(std::string& out_buffer, const std::string& filePath) const {
+    namespace FS = std::filesystem;
+    FS::path p(filePath);
+    p.make_preferred();
+    bool path_is_directory = FS::is_directory(p);
+    bool path_not_exist = !FS::exists(p);
+    bool not_valid_path = path_is_directory || path_not_exist;
+    if(not_valid_path) {
+        return false;
+    }
+
+    std::ifstream ifs{ p };
+    //Dump ifstream buffer directly into stringstream and convert to string
+    out_buffer = std::string(static_cast<const std::stringstream&>(std::stringstream() << ifs.rdbuf()).str());
+    return true;
+}
+
+//TODO: Needs Linux/MacOSX implementations!
```

Comment:
The Linux and MacOSX implementations of GetExePath? Yeah, I did not know the OS-specific APIs and intended to update them later.

In plain C++, Generally the working path where the executable is run from is the first argument from the command line (argv[0]), the default std::filesystem::current_path() (not calling it with arguments first) returns the working directory which is the same as argv[0] but both of these can be changed via $PATH env_vars and calls to std::filesystem::current_path(std::filesystem::path new_path).

A quick google search netted an informative StackOverflow answer:
https://stackoverflow.com/a/1024933/421178
and another posted links to the required OS documentation:
https://stackoverflow.com/a/1024937/421178

It looks like the MacOSX implementation is as simple as the Windows one:
 - uint32_t size = 0;
 - std::string path{};
 - Call _NSGetExecutablePath(path.data(), &size) to get the required size
 - Resize the buffer: path.resize(size)
 - Call _NSGetExecutablePath(path.data(), &size) again and cram it into a buffer.
 - Construct a std::filesystem::path from the buffer and convert it to a preferred canonical path.
 - std::filesystem::path p{path};
 - p = p.canonical();
 - p.make_preferred();

Linux...not so much. Create a path to /proc/self/exe which is a symbolic link then convert it to an absolute canonical preferred path...

Reply:
One thing that is definitely a problem is the build script is trying to build the vc14 project instead of the vc17 project. std::filesystem did not exist in vc14. As usual I forgot to update the AppVeyor.yaml script. :/

Reply:
@DanRStevens GetExePath now calls correct OS-dependent functions. I cannot locally test if any of it is logically correct, only syntactically through the TravisCI, as I do not have a local Linux or Mac OSX available.

Reply:
Yes, the CI builds have only been doing straight builds so far. We are only just now starting to add post-build unit tests. We would need unit tests for this to check if the code runs correctly.

Reply:
@DanRStevens  The linux builds pass. After several rounds of testing and googling, I've determined TravisCI's macOSX gcc, llvm and clang versions support neither the <experimental/filesystem> nor the <filesystem> headers AT ALL.

Reply:
That was my finding about a year ago on another project as well. That other project never ended up officially supporting MacOS because of that. A decision to be revisited when tools have a chance to catch up.